File: src/dataset/cpp/protein-translation_exercism-tqa236/claude/ClaudeSonnet4_protein_translation_v3.cpp
#include "protein_translation.h"

#include <unordered_map>
#include <string_view>

namespace protein_translation {

const std::unordered_map<std::string_view, std::string_view> translator = {
    {"AUG", "Methionine"}, {"UUU", "Phenylalanine"}, {"UUC", "Phenylalanine"},
    {"UUA", "Leucine"},    {"UUG", "Leucine"},       {"UCU", "Serine"},
    {"UCC", "Serine"},     {"UCA", "Serine"},        {"UCG", "Serine"},
    {"UAU", "Tyrosine"},   {"UAC", "Tyrosine"},      {"UGU", "Cysteine"},
    {"UGC", "Cysteine"},   {"UGG", "Tryptophan"},    {"UAA", "STOP"},
    {"UAG", "STOP"},       {"UGA", "STOP"}};

std::vector<codon> proteins(const ARN& arn) {
    if (arn.empty()) return {};
    
    std::vector<polypeptide> protein_names;
    protein_names.reserve(arn.length() / 3);
    
    for (size_t i = 0; i + 2 < arn.length(); i += 3) {
        std::string_view codon_view = std::string_view(arn).substr(i, 3);
        auto it = translator.find(codon_view);
        if (it == translator.end()) continue;
        
        if (it->second == "STOP") break;
        protein_names.emplace_back(it->second);
    }
    
    return protein_names;
}

}  // namespace protein_translation

File: src/dataset/javascript/protein-translation_exercism-ThomasZumsteg/claude/ClaudeSonnet4_protein-translation_v3.js
const translate = new Map([
    ['AUG', 'Methionine'],
    ['UCA', 'Serine'],
    ['UCC', 'Serine'],
    ['UCG', 'Serine'],
    ['UCU', 'Serine'],
    ['UUA', 'Leucine'], 
    ['UUC', 'Phenylalanine'],
    ['UUG', 'Leucine'],
    ['UUU', 'Phenylalanine'],
    ['UAU', 'Tyrosine'],
    ['UAC', 'Tyrosine'],
    ['UGC', 'Cysteine'],
    ['UGU', 'Cysteine'],
    ['UGG', 'Tryptophan'],
    ['UAA', 'STOP'],
    ['UAG', 'STOP'],
    ['UGA', 'STOP']
]);

module.exports = function(rna) {
    if (!rna) return [];
    
    const result = [];
    const length = rna.length;
    
    for (let i = 0; i < length; i += 3) {
        const codon = rna.substr(i, 3);
        if (codon.length < 3) break;
        
        const protein = translate.get(codon);
        if (protein === 'STOP') break;
        if (protein === undefined) throw new Error('Invalid codon');
        
        result.push(protein);
    }
    
    return result;
};